once/_iterator_wrappers.py

```diff
@@ -24,116 +25,156 @@ def __init__(self) -> None:
         self.exception: Exception | None = None
         self.finished = False
 
+    def fast_path(self):
+        """Should only be computed with a lock."""
+        return self.finished and self.exception is None
 
-# TODO(matt): Refactor AsyncGeneratorWrapper to use state enums, and add error-handling.
-class AsyncGeneratorWrapper:
+
+class _IteratorAction(enum.Enum):
+    # Generating the next value from the underlying iterator
+    GENERATING = 1
+    # Yield an already computed value
+    YIELDING = 2
+    # Waiting for the underlying iterator, already triggered from another call.
+    WAITING = 3
+    # We can return, we are done!
+    RETURNING = 4
+
+
+class _GeneratorWrapperBase:
+    """Base class for generator wrapper.
+
+    Even though the class stores a result, all of the methods separately take a result input.
+    Why is that? Great question.
+    During a call to yield_results, we should grab a reference to the existing self.result, and
```

Comment:
I appreciate the detailed comment, but I struggled to understand it until our slack chat. Thoughts on:

```python
While yielding results simultaneously from multiple tasks or threads, we want to support configurable
outcomes for exception handling.
Specifically, taking `result` as an argument is critical in order to support any *already executing*
iterators finishing with their cached result and exception, but restarting new iterators from the beginning.
This is done by creating a local copy of `result` when the iterator is started and passing that around thereafter.

TODO(ali and matt): Consider whether we want to start from the beginning, or just from the exception,
for non-cached exception configurations.
```

Reply:
added a detailed comment